spec/features/users_spec.rb

```diff
@@ -0,0 +1,148 @@
+      visit new_user_path
+
+      within "form" do
+        fill_in "Name", with: "Mary"
+        fill_in "Password", with: "secret"
+        click_button "Sign up"
+      end
+
+      expect(page).to have_text("User could not be created")
+      expect(page).to have_text("Please fill in the form to sign up")
+    end
+
+    scenario "without a password" do
+      visit new_user_path
+
+      within "form" do
+        fill_in "Name", with: "Mary"
+        fill_in "Email", with: "Mary@example.com"
+        click_button "Sign up"
+      end
+
+      expect(page).to have_text("User could not be created")
+      expect(page).to have_text("Please fill in the form to sign up")
+    end
+  end
+
+  feature "user can see their show page" do
+    scenario "signing up with valid attributes" do
+      user = create_user("John", "John@example.com", "password")
+      visit user_path(user)
+
+      expect(page).to have_text("Welcome John")
+    end
+  end
+
+  feature "user can visit the user edit page" do
+    scenario "logging in with valid attributes" do
+      user = create_user("John", "John@example.com", "password")
+      visit edit_user_path(user)
+
+      expect(page).to have_text("Please fill in the form to edit your profile")
+    end
+  end
+
+  feature "user can update their profile" do
+    scenario "logging in with valid attributes" do
+      user = create_user("John", "John@example.com", "password")
+      visit edit_user_path(user)
+
+      within "form" do
+        fill_in "Email", with: "JT@example.com"
+        click_button "Edit profile"
+      end
+
+      expect(page).to have_text("Your email address is: JT@example.com")
+    end
+  end
+
+  feature "user can't update their profile" do
+    scenario "without entering a name" do
+      user = create_user("John", "John@example.com", "password")
+      visit edit_user_path(user)
+
+      within "form" do
+        fill_in "Name", with: ""
+        fill_in "Email", with: "JT@example.com"
+        click_button "Edit profile"
+      end
+
+      expect(page).to have_text("User could not be updated.")
+      expect(page).to have_text("Please fill in the form to edit your profile")
+    end
+
+    scenario "without entering an email" do
+      user = create_user("John", "John@example.com", "password")
+      visit edit_user_path(user)
+
+      within "form" do
+        fill_in "Name", with: "John"
+        fill_in "Email", with: ""
+        click_button "Edit profile"
+      end
+
+      expect(page).to have_text("User could not be updated.")
+      expect(page).to have_text("Please fill in the form to edit your profile")
+    end
+  end
+
+  feature "user can delete their account" do
+    scenario "with valid attributes" do
+      user = create_user("John", "John@example.com", "password")
+      visit edit_user_path(user)
+      click_link "Delete Account"
+
+      expect(page).to have_text("Welcome to Budget App")
+    end
+  end
+
+  def create_user(name, email, password)
```

Comment:
This method is not really needed, if you use factories

Reply:
Thanks Elle, I will remove it!